src/changeset.py

```diff
@@ -0,0 +1,201 @@
+from __future__ import unicode_literals
+
+import itertools
+
+from models import (
+    Relation,
+    UnitPlacement,
+)
+
+
+def _parse_v3_unit_placement(placement):
+    """Return a UnitPlacement for bundles version 3, given a placement string.
+    """
+    container = machine = service = unit = ''
+    if ':' in placement:
+        container, placement = placement.split(':')
+    if '=' in placement:
+        placement, unit = placement.split('=')
+    if placement.isdigit():
+        machine = placement
+    else:
+        service = placement
+    return UnitPlacement(container, machine, service, unit)
+
+
+def _parse_v4_unit_placement(placement):
+    """Return a UnitPlacement for bundles version 4, given a placement string.
+    """
+    container = machine = service = unit = ''
+    if ':' in placement:
+        container, placement = placement.split(':')
+    if '/' in placement:
+        placement, unit = placement.split('/')
+    if placement.isdigit():
+        machine = placement
+    else:
+        service = placement
+    return UnitPlacement(container, machine, service, unit)
+
+
+class ChangeSet(object):
+    """Hold the state for parser handlers.
+
+    Also expose methods to send and receive changes (usually Python dicts).
+    """
+    services_added = {}
+    machines_added = {}
+
+    def __init__(self, bundle):
+        self.bundle = bundle
+        self._changeset = []
+        self._counter = itertools.count()
+
+    def send(self, change):
+        """Store a change in this change set."""
+        self._changeset.append(change)
+
+    def recv(self):
+        """Return all the collected changes.
+
+        Changes are stored using self.send().
+        """
+        changeset = self._changeset
+        self._changeset = []
+        return changeset
+
+    def next_action(self):
+        """Return an incremental integer to be included in the changes ids."""
+        return next(self._counter)
+
+
+def parse(bundle, handler=None):
+    """Return a generator yielding changes required to deploy the given bundle.
+
+    The bundle argument is a YAML decoded Python dict.
+    """
+    changeset = ChangeSet(bundle)
+    if handler is None:
+        handler = handle_services
```

Comment:
Is there any reason to not just have handle_services passed in as the default for `handler`, rather than None? Then you can just remove this check.